Simulator/Vertices/EventBuffer.cpp

```diff
@@ -42,27 +41,16 @@ const string &EventBuffer::getDataType() const
    return basicDataType_;
 }
 
-void EventBuffer::startNewEpoch()
-{
-   epochStart_ = queueEnd_;
-   numEventsInEpoch_ = 0;
-}
-
 int EventBuffer::getNumElements() const
-{
-   return getNumEventsInEpoch();
-}
-
-int EventBuffer::getNumEventsInEpoch() const
 {
    return numEventsInEpoch_;
 }
```

Comment:
I see. Indeed there is a little bit distinction here. I will fix it.

Reply:
This is a bit of confusing for me too.
 In the RecordableVector class, the number of element in the vector is the number of elements in the current epoch but in the EventBuffer class, the number of elements in the vector not equals to the number of elements in the current epoch and the Recorder only wants to get the number of elements in each epoch from the EventBuffer. This is the reason why I decide to merge it and overlook the distinction. We can have two virtual methods size() and getNumElementInEpoch() in both RecordableVector and EventBuffer. Does this make sense?

Reply:
OK, thinking further. A `RecordableVector` is just an ordinary-appearing vector, but with the extras needed to make it `Recordable` (regardless of what its parent class actually is). Any class that uses it manipulates it like a `vector` with those additional `Recordable` methods (the care simulator doesn't use those extra methods, of course, so it just is used like a `vector` there).

On the other hand, an `EventBuffer` has a second interface: to the `Recorder` classes, it looks like a `RecordableVector`, but to the core simulator, it looks like a buffer: it can be written to and past elements can be retrieved. So, the current `getNumEventsInEpoch()` method is the `vector` size operation. But, there is a difference with `RecordableVector`, in that after the `Recorder` is done with it and "clears" it, the buffer state isn't modified — it looks like it's empty when you think of it as a `vector`, but nothing has changed when you think of it as a buffer. So, there is a need to maintain the buffer state information, which means we need to know where the most recent event is and we need to know how many events are in the buffer. That's because, until the buffer fills up, wraps around, and new events start over-writing the oldest events, the number of events in the buffer is smaller than the buffer capacity. Of course, once it wraps around, the number of events stops increasing.

So, maybe we don't need a method to return the number of elements in the buffer; we just need a member variable.

Reply:
The Recorder needs to get the number of elements in the RecordableVector object(the size of vector) while it needs to get the number of event in the current epoch from the EventBuffer. If we only have one such method in common interface, the RecordableBase, maybe we can document them in the comment. For example
RecordableBase{
    // get the number of element that need to be record in the object
    virtual int getNumElements() constant;
} 
RecordableVector: public RecordableBase{
    virtual int getNumElement() constant override
}
EventBuffer: public RecordableVector{
    //we need this virtual method because it is called by the Recorder
    virtual int getNumElement() constant override{
       return getNumElementInEpoch or numEventsInEpoch;
    }
    //we still need this method because it is called by variable owner class such as AllSpikingNeurons
    int getNumElementInEpoch() constant{
          return numEventsInEpoch
    }
private: 
    numEventsInEpoch_// there already is a data member showing showing the events information, maybe I need to rename it as numElementsInEpoch.
}

Reply:
I will rename the data member numEventsInEpoch_ as numElementsInEpoch_

Reply:
Does the variable owner class need to call `EventBuffer:: getNumElementInEpoch()`, or is that just needed within the `EventBuffer` class? If the latter, then we don't need that accessor; we can just use the data member.

Reply:
The variable owner class, AllSpikingNeurons, needs to call this method.
```c++
//data member
vector<EventBuffer> vertexEvent_
// in cpp file, 
 vertexEvent[I].getNumElementInEpoch()) 
```
So I think we still need to keep it.